server/src/controllers/course.ts

```diff
@@ -106,12 +109,130 @@ export async function addCourse(req: Request, res: Response): Promise<void> {
   }
 }
 
+enum GradingType {
+  PassFail = 'PASSFAIL',
+  Numerical = 'NUMERICAL'
+}
+
+enum Period {
+  I = 'I',
+  II = 'II',
+  III = 'III',
+  IV = 'IV',
+  V = 'V'
+}
+
+enum TeachingMethod {
+  Lecture = 'LECTURE',
+  Exam = 'EXAM'
+}
+
+interface CourseInstanceAddRequest {
+  gradingType: GradingType;
+  startingPeriod: Period;
+  endingPeriod: Period;
+  teachingMethod: TeachingMethod;
+  responsibleTeacher: number;
+  startDate: Date;
+  endDate: Date;
+}
+
+const courseInstanceAddRequestSchema: yup.AnyObjectSchema = yup.object().shape({
+  gradingType: yup
+    .string()
+    .oneOf([GradingType.PassFail, GradingType.Numerical])
+    .required(),
+  startingPeriod: yup
+    .string()
+    .oneOf([Period.I, Period.II, Period.III, Period.IV, Period.V])
+    .required(),
+  endingPeriod: yup
+    .string()
+    .oneOf([Period.I, Period.II, Period.III, Period.IV, Period.V])
+    .required(),
+  teachingMethod: yup
+    .string()
+    .oneOf([TeachingMethod.Lecture, TeachingMethod.Exam])
+    .required(),
+  responsibleTeacher: yup
+    .number()
+    .required(),
+  startDate: yup
+    .date()
+    .required(),
+  endDate: yup
+    .date()
+    .required()
+});
+
+export async function addCourseInstance(req: Request, res: Response): Promise<Response> {
+  try {
+    const courseId: number = Number(req.params.courseId);
+
+    await courseInstanceAddRequestSchema.validate(req.body, { abortEarly: false });
+
+    const request: CourseInstanceAddRequest = req.body;
+
+    const course: Course | null = await models.Course.findOne({
+      where: {
+        id: courseId
+      },
+    });
+
+    if (course == null) {
+      throw new Error(`Course with ID ${courseId} does not exist`);
+    }
+
+    const teacher: User | null = await models.User.findOne({
+      where: {
+        id: request.responsibleTeacher
+      },
+    });
+
+    // TODO: Also check that user has correct role
+    if (teacher == null) {
+      throw new Error(`Teacher with ID ${request.responsibleTeacher} does not exist`);
+    }
+
+    const newInstance: CourseInstance = await models.CourseInstance.create({
+      courseId: courseId,
+      gradingType: request.gradingType,
+      startingPeriod: request.startingPeriod,
+      endingPeriod: request.endingPeriod,
+      teachingMethod: request.teachingMethod,
+      responsibleTeacher: request.responsibleTeacher,
+      startDate: request.startDate,
+      endDate: request.endDate
+    });
+
+    return res.send({
+      success: true,
+      instance: newInstance
+    });
+  } catch (error) {
+
+    if (error instanceof yup.ValidationError) {
+      res.status(400);
+      return res.send({
+        success: false,
+        error: error.errors
+      });
+    }
+
+    res.status(401);
```

Comment:
Maybe we could consider returning a 404 for some of the errors, if they point to a course/teacher not having been found. A 500 could also be suitable for database errors.

Reply:
You're right, the status codes used here are not ideal. @jyrge also commented on this, and we thought that it would be best to discuss our API specs and error handling at the beginning of the next sprint and then go back to update our old routes. So far we've all been doing things in slightly different ways, so getting some collective consensus going forward is desirable.

Reply:
+1 for the consensus. With that, this code should be good for now.